pkg/pool/metric_pool.go

```diff
@@ -0,0 +1,50 @@
+package pool
+
+import (
+	"sync"
+
+	"github.com/atlassian/gostatsd"
+)
+
+// MetricPool is a strongly typed wrapper around a sync.Pool for *gostatsd.Metric, it provides
+// two main benefits: 1) metrics are very short lived and we create a lot of them, 2) reuse
+// of the tags buffer
+type MetricPool struct {
+	p             sync.Pool
+	estimatedTags int
+}
+
+// NewMetricPool returns a new metric pool.
+func NewMetricPool(estimatedTags int) *MetricPool {
+	return &MetricPool{
+		p: sync.Pool{
+			New: func() interface{} {
+				return &gostatsd.Metric{}
+			},
+		},
+		estimatedTags: estimatedTags,
+	}
+}
+
+// Get returns a *gostatsd.Metric suitable for holding a metric.  The DoneFunc should be called
+// when the metric is no longer required.  It must not be called earlier, and the Tags field may
+// be reused.
+func (mp *MetricPool) Get() *gostatsd.Metric {
+	m := mp.p.Get().(*gostatsd.Metric)
+	if m.DoneFunc != nil { // it was re-used, and the data needs cleaning
```

Comment:
This all looks a bit messy. What if all the field initialisation went into the Metric.Reset() and this is called from DoneFunc before returning the Metric to the MetricPool?

Reply:
Doing it before returning may be wasted work, because the pool is flushed every GC.  I like pushing it to another function though because it puts it near the type declaration.